src/backend/core/api/viewsets.py

```diff
@@ -331,30 +321,68 @@ class DocumentViewSet(
     queryset = models.Document.objects.all()
     ordering = ["-updated_at"]
 
-    def perform_create(self, serializer):
+    def list(self, request, *args, **kwargs):
+        """Restrict resources returned by the list endpoint"""
+        queryset = self.filter_queryset(self.get_queryset())
+        user = self.request.user
+        if user.is_authenticated:
+            queryset = queryset.filter(
+                Q(accesses__user=user)
+                | Q(accesses__team__in=user.teams)
+                | (
+                    Q(link_traces__user=user)
+                    & ~Q(link_reach=models.LinkReachChoices.RESTRICTED)
+                )
+            )
+        else:
+            queryset = queryset.none()
+
+        page = self.paginate_queryset(queryset)
+        if page is not None:
+            serializer = self.get_serializer(page, many=True)
+            return self.get_paginated_response(serializer.data)
+
+        serializer = self.get_serializer(queryset, many=True)
+        return drf_response.Response(serializer.data)
+
+    def retrieve(self, request, *args, **kwargs):
         """
-        Override perform_create to use the provided ID in the payload if it exists
+        Add a trace that the document was accessed by a user. This is used to list documents
+        on a user's list view even though the user has no specific role in the document (link
+        access when the link reach configuration of the document allows it).
         """
-        document_id = self.request.data.get("id")
-        document = serializer.save(id=document_id) if document_id else serializer.save()
+        instance = self.get_object()
+        serializer = self.get_serializer(instance)
+
+        if self.request.user.is_authenticated:
+            try:
+                # Add a trace that the user visited the document (this is needed to include
+                # the document in the user's list view)
+                models.LinkTrace.objects.create(
+                    document=instance,
+                    user=self.request.user,
+                )
```

Comment:
It creates an error if you access the document twice and more, because of the unicity document/user.

Reply:
hmmm. The try/except is here for that? I didn't manage to reproduce the error. Can you introspect what are the values of instance and self.request.user when this occurs and what is the error type generated?

Reply:
I get a 400, "Document/user link trace with this User and Document already exists.".

![image](https://github.com/user-attachments/assets/9482b1a9-f80d-4d16-94e4-67f97cc93206)

If a change the method to `update_or_create`, it is working. 
```py
models.LinkTrace.objects.update_or_create
```

Reply:
Ok, it is because of the type of error, it is a `ValidationError` and not a `IntegrityError`
https://github.com/numerique-gouv/impress/blob/cec2531352184163a0fc1988deb829c0222c2aaa/src/backend/core/api/viewsets.py#L365